pyodide-build/pyodide_build/out_of_tree/pypi.py

```diff
@@ -0,0 +1,386 @@
+        with tempfile.TemporaryFile(mode="w+") as f:
+            try:
+                with stream_redirected(to=f, stream=sys.stdout):
+                    with stream_redirected(to=f, stream=sys.stderr):
+                        wheel_path = build.run(
+                            PyPIProvider.BUILD_EXPORTS, PyPIProvider.BUILD_FLAGS
+                        )
+            except BaseException as e:
+                print(" Failed\n Error is:")
+                f.seek(0)
+                sys.stdout.write(f.read())
+                raise e
+
+        OKGREEN = "\033[92m"
+        ENDC = "\033[0m"
+        print(OKGREEN, "Success", ENDC)
+
+        cache_entry["path"] = wheel_path
+        BUILD_CACHE[url] = cache_entry
+        return wheel_path
+
+
+class Candidate:
+    def __init__(self, name, version, url=None, extras=None):
+        self.name = canonicalize_name(name)
+        self.version = version
+        self.url = url
+        self.extras = extras
+
+        self._metadata = None
+        self._dependencies = None
+
+    def __repr__(self):
+        if not self.extras:
+            return f"<{self.name}=={self.version}>"
+        return f"<{self.name}[{','.join(self.extras)}]=={self.version}>"
+
+    @property
+    def metadata(self):
+        if self._metadata is None:
+            self._metadata = get_metadata_for_wheel(self.url)
+        return self._metadata
+
+    @property
+    def requires_python(self):
+        return self.metadata.get("Requires-Python")
+
+    def _get_dependencies(self):
+        deps = self.metadata.get_all("Requires-Dist", [])
+        extras = self.extras if self.extras else [""]
+
+        for d in deps:
+            r = Requirement(d)
+            if r.marker is None:
+                yield r
+            else:
+                for e in extras:
+                    if r.marker.evaluate({"extra": e}):
+                        yield r
+
+    @property
+    def dependencies(self):
+        if self._dependencies is None:
+            self._dependencies = list(self._get_dependencies())
+        return self._dependencies
+
+
+if TYPE_CHECKING:
+    APBase = AbstractProvider[Requirement, Candidate, str]
+else:
+    APBase = AbstractProvider
+
+
+class ExtrasProvider(APBase):
+    """A provider that handles extras."""
+
+    def get_extras_for(self, requirement_or_candidate):
+        """Given a requirement or candidate, return its extras.
+        The extras should be a hashable value.
+        """
+        raise NotImplementedError
+
+    def get_base_requirement(self, candidate):
+        """Given a candidate, return a requirement that specifies that
+        project/version.
+        """
+        raise NotImplementedError
+
+    def identify(self, requirement_or_candidate):
+        base = super().identify(requirement_or_candidate)
+        return base
+
+    def get_dependencies(self, candidate):
+        deps = list(super().get_dependencies(candidate))
```

Comment:
Since `AbstractProvider.get_dependencies` raises `NotImplementedError`
https://github.com/sarugaku/resolvelib/blob/master/src/resolvelib/providers.py#L109
doesn't this just unconditionally raise `NotImplementedError`? I think you should get rid of `ExtrasProvider` and combine the logic here with the logic in `PyPIProvider.get_dependencies`.